File: Euclid/Game/Room/Entity/RoomEntity.cs
using Euclid.Game.Entity.Humanoid;
using Euclid.Messages.Outgoing;
using Euclid.Util.Extensions;
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Linq;

namespace Euclid.Game
{
    public abstract class RoomEntity
    {
        public Humanoid Entity { get; set; }
        public Room Room { get; set; }
        public Position Position { get; set; }
        public Position Next { get; set; }
        public Position Goal { get; set; }
        public List<Position> PathList { get; private set; }
        public bool IsWalking { get; set; }
        public int InstanceId { get; set; }
        public bool NeedsUpdate { get; set; }
        public int RoomId => Room != null ? Room.Data.Id : 0;
        public RoomTile CurrentTile => Position != null ? (Position.GetTile(Room) ?? null) : null;
        public Item CurrentItem => CurrentTile?.HighestItem;
        public RoomTimerManager TimerManager { get; set; }
        public string AuthenticateTeleporterId { get; set; }
        public int AuthenticateRoomId { get; set; }
        public bool WalkingAllowed { get; internal set; }
        public virtual ITaskObject TaskObject { get; set; }
        public int DanceId { get; set; }
        public bool IsDancing => DanceId > 0;
        public bool HasEffect => EffectId > 0;
        public bool IsSitting => Status.ContainsKey("sit");
        public int EffectId { get; set; }
        public bool BeingKicked { get; set; }

        /// <summary>
        /// Get the status handling, the value is the value string and the time it was added.
        /// </summary>
        public ConcurrentDictionary<string, RoomUserStatus> Status { get; set; }

        public RoomEntity(Humanoid entity)
        {
            Entity = entity;
            TimerManager = new RoomTimerManager();
            AuthenticateTeleporterId = null;
        }

        /// <summary>
        /// Reset handler called when user leaves room
        /// </summary>
        public void Reset()
        {
            Status = new ConcurrentDictionary<string, RoomUserStatus>();
            IsWalking = false;
            Goal = null;
            Next = null;
            InstanceId = -1;
            DanceId = 0;
            Room = null;
            TimerManager.Reset();
            WalkingAllowed = true;
        }

        /// <summary>
        /// Chat message handling
        /// </summary>
        public void Talk(ChatMessageType chatMessageType, string chatMsg, List<Player> recieveMessages = null)
        {
            if (recieveMessages == null)
                recieveMessages = Room.EntityManager.GetEntities<Player>();

            recieveMessages = recieveMessages.Distinct().ToList();

            // Send talk message to room
            foreach (Player player in recieveMessages)
            {
                switch (chatMessageType)
                {
                    case ChatMessageType.CHAT:
                        player.Send(new CHAT(this.Entity.EntityData.Name, chatMsg));
                        break;
                    case ChatMessageType.SHOUT:
                        player.Send(new SHOUT(this.Entity.EntityData.Name, chatMsg));
                        break;
                    case ChatMessageType.WHISPER:
                        player.Send(new WHISPER(this.Entity.EntityData.Name, chatMsg));
                        break;
                }
            }
        }

        /// <summary>
        /// Get chat gesture for message
        /// </summary>
        private int GetChatGesture(string chatMsg)
        {
            chatMsg = chatMsg.ToLower();

            if (chatMsg.Contains(":)") || chatMsg.Contains(":d") || chatMsg.Contains("=]") ||
                chatMsg.Contains("=d") || chatMsg.Contains(":>"))
            {
                return 1;
            }

            if (chatMsg.Contains(">:(") || chatMsg.Contains(":@"))
                return 2;

            if (chatMsg.Contains(":o"))
                return 3;

            if (chatMsg.Contains(":(") || chatMsg.Contains("=[") || chatMsg.Contains(":'(") || chatMsg.Contains("='["))
                return 4;

            return 0;
        }

        /// <summary>
        /// Request move handler
        /// </summary>
        /// <param name="x">x coord goal</param>
        /// <param name="y">y coord goal</param>
        public void Move(int x, int y)
        {
            if (Room == null)
                return;

            if (Next != null)
            {
                var oldPosition = Next.Copy();
                Position.X = oldPosition.X;
                Position.Y = oldPosition.Y;
                Position.Z = Room.Model.TileHeights[oldPosition.X, oldPosition.Y];
                NeedsUpdate = true;
            }

            Goal = new Position(x, y);

            var goalTile = Goal.GetTile(Room);

            if (!RoomTile.IsValidTile(Room, Entity, Goal))
                return;

            var pathList = Pathfinder.FindPath(Entity, Room, Position, Goal);

            if (pathList == null || !pathList.Any())
                return;

            PathList = pathList;
            IsWalking = true;
        }

        /// <summary>
        /// Kick a user from the room.
        /// </summary>

        public void kick(bool allowWalking = true)
        {
            this.BeingKicked = true;
            Position doorLocation = this.Room.Model.Door;


            // If we're standing in the door, walk to room
            if (this.Position == doorLocation)
            {
                this.Room.EntityManager.LeaveRoom(this.Entity);
                return;
            }

            // Attempt to walk to the door
            this.Move(doorLocation.X, doorLocation.Y);
            this.WalkingAllowed = allowWalking;

            // If user isn't walking, leave immediately
            if (!this.IsWalking)
            {
                this.Room.EntityManager.LeaveRoom(this.Entity);
            }
        }

        /// <summary>
        /// Stopped walking handler
        /// </summary>
        public void StopWalking()
        {
            if (!this.IsWalking)
                return;

            this.IsWalking = false;
            this.PathList.Clear();
            this.Next = null;
            this.RemoveStatus("mv");
            this.InteractItem();
            this.NeedsUpdate = true;

            bool leaveRoom = this.BeingKicked;
            Position doorPosition = this.Room.Model.Door;//.getModel().getDoorLocation();

            if (doorPosition == this.Position)
            {
                leaveRoom = true;
            }

            if (this.Room.Data.IsPublicRoom)
            {
                /*if (WalkwaysManager.getInstance().getWalkway(this.getRoom(), doorPosition) != null)
                {
                    leaveRoom = false;
                }*/
            }

            // Leave room if the tile is the door and we are in a flat or we're being kicked
            if (leaveRoom || this.BeingKicked)
            {
                this.Room.EntityManager.LeaveRoom(this.Entity, true);
                return;
            }
        }

        /// <summary>
        /// Interact with current item by calling entity stop on interactor
        /// </summary>
        public void InteractItem()
        {
            var roomTile = CurrentTile;

            if (roomTile == null)
                return;

            Position.Z = roomTile.GetWalkingHeight();

            Item item = CurrentItem;

            if (item == null || (
                !item.Definition.Data.IsChair ||
                !item.Definition.Data.IsBed))
            {
                if (ContainsStatus("sit") || ContainsStatus("lay"))
                {
                    RemoveStatus("sit");
                    RemoveStatus("lay");
                }
            }

            if (item != null && Entity is Humanoid humanoid)
            {
                if (item.Definition.Data.IsChair)
                {
                    humanoid.RoomEntity.Position.Rotation = item.Position.Rotation;
                    humanoid.RoomEntity.AddStatus("sit", Convert.ToString((int)item.Height));
                    humanoid.RoomEntity.NeedsUpdate = true;
                }

                if (item.Definition.Data.IsBed)
                {
                    humanoid.RoomEntity.Position.Rotation = item.Position.Rotation;
                    humanoid.RoomEntity.AddStatus("lay", Convert.ToString((int)item.Height));
                    humanoid.RoomEntity.NeedsUpdate = true;
                }
                //item.Interactor.OnStop(this.Entity);
            }

            this.NeedsUpdate = true;
        }

        /// <summary>
        /// Get if entity contains status
        /// </summary>
        public bool ContainsStatus(string statusKey)
        {
            return Status.ContainsKey(statusKey);
        }

        /// <summary>
        /// Adds a status with a key and value, along with the int64 time of when the status was added.
        /// </summary>
        /// <param name="key">the key</param>
        /// <param name="value">the value</param>
        public void AddStatus(string key, string value)
        {
            this.RemoveStatus(key);
            Status.TryAdd(key, new RoomUserStatus(key, value));
        }

        /// <summary>
        /// Removes a status by its given key
        /// </summary>
        /// <param name="key">the key to check for</param>
        public void RemoveStatus(string key)
        {
            if (Status.ContainsKey(key))
                this.Status.Remove(key);
        }

        /// <summary>
        /// Warp etntiy to position
        /// </summary>
        /// <param name="id"></param>
        /// <param name="v"></param>
        public void Warp(Position targetPosition, bool instantUpdate = false)
        {
            RoomTile oldTile = CurrentTile;

            if (oldTile != null)
            {
                oldTile.RemoveEntity(Entity);
            }

            if (Next != null)
            {
                RoomTile nextTile = Next.GetTile(Room);

                if (nextTile != null)
                {
                    nextTile.RemoveEntity(Entity);
                }
            }

            Position = targetPosition.Copy();
            RefreshHeight(targetPosition);

            RoomTile newTile = CurrentTile;

            if (newTile != null)
            {
                newTile.AddEntity(Entity);
            }

            if (instantUpdate && Room != null)
            {
                InteractItem();
                //Room.Send(new UsersStatusComposer(List.Create(Entity)));
            }
        }

        /// <summary>
        /// Refresh height of entity
        /// </summary>
        private void RefreshHeight(Position newPosition)
        {
            var targetPosition = newPosition ?? Position;

            var oldTile = Position.GetTile(Room);
            var newTile = targetPosition.GetTile(Room);

            if (oldTile.GetWalkingHeight() != newTile.GetWalkingHeight())
            {
                Position.Z = newTile.GetWalkingHeight();
                NeedsUpdate = true;
            }
        }
    }

}

File: euclid/src/main/java/euclid/game/entity/RoomEntity.java
package euclid.game.entity;

import euclid.game.Player;
import euclid.game.item.Item;
import euclid.game.pathfinder.Pathfinder;
import euclid.game.pathfinder.Position;
import euclid.game.pathfinder.Rotation;
import euclid.game.room.Room;
import euclid.game.room.RoomTile;
import euclid.game.room.RoomTimerManager;
import euclid.game.room.tasks.ITaskObject;
import euclid.game.util.ChatMessageType;
import euclid.messages.outgoing.CHAT;
import euclid.messages.outgoing.SHOUT;
import euclid.messages.outgoing.WHISPER;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class RoomEntity {

    private Humanoid entity;
    private Room room;
    private Position position;
    private Position next;
    private Position goal;
    private List<Position> pathList;
    private boolean isWalking;
    private int instanceId;
    private boolean needsUpdate;
    private RoomTimerManager timerManager;
    private String authenticateTeleporterId;
    private int authenticateRoomId;
    private boolean walkingAllowed = true;
    private ITaskObject taskObject;
    private int danceId;
    private int effectId;
    private boolean beingKicked;
    private Map<String, EntityStatus> statuses;

    public RoomEntity(Humanoid entity) {
        this.entity = entity;
        this.timerManager = new RoomTimerManager();
        this.authenticateTeleporterId = null;
        this.statuses = new HashMap<>();
        this.pathList = new ArrayList<>();
    }

    public Humanoid getEntity() {
        return entity;
    }

    public void setEntity(Humanoid entity) {
        this.entity = entity;
    }

    public Room getRoom() {
        return room;
    }

    public void setRoom(Room room) {
        this.room = room;
    }

    public Position getPosition() {
        return position;
    }

    public void setPosition(Position position) {
        this.position = position;
    }

    public Position getNext() {
        return next;
    }

    public void setNext(Position next) {
        this.next = next;
    }

    public Position getGoal() {
        return goal;
    }

    public void setGoal(Position goal) {
        this.goal = goal;
    }

    public List<Position> getPathList() {
        return pathList;
    }

    public void setPathList(List<Position> pathList) {
        this.pathList = pathList;
    }

    public boolean isWalking() {
        return isWalking;
    }

    public void setWalking(boolean walking) {
        isWalking = walking;
    }

    public int getInstanceId() {
        return instanceId;
    }

    public void setInstanceId(int instanceId) {
        this.instanceId = instanceId;
    }

    public boolean isNeedsUpdate() {
        return needsUpdate;
    }

    public void setNeedsUpdate(boolean needsUpdate) {
        this.needsUpdate = needsUpdate;
    }

    public int getRoomId() {
        return room != null ? room.getData().getId() : 0;
    }

    public RoomTile getCurrentTile() {
        return position != null ? position.getTile(room) : null;
    }

    public Item getCurrentItem() {
        RoomTile tile = getCurrentTile();
        return tile != null ? tile.getHighestItem() : null;
    }

    public RoomTimerManager getTimerManager() {
        return timerManager;
    }

    public void setTimerManager(RoomTimerManager timerManager) {
        this.timerManager = timerManager;
    }

    public String getAuthenticateTeleporterId() {
        return authenticateTeleporterId;
    }

    public void setAuthenticateTeleporterId(String authenticateTeleporterId) {
        this.authenticateTeleporterId = authenticateTeleporterId;
    }

    public int getAuthenticateRoomId() {
        return authenticateRoomId;
    }

    public void setAuthenticateRoomId(int authenticateRoomId) {
        this.authenticateRoomId = authenticateRoomId;
    }

    public boolean isWalkingAllowed() {
        return walkingAllowed;
    }

    public void setWalkingAllowed(boolean walkingAllowed) {
        this.walkingAllowed = walkingAllowed;
    }

    public ITaskObject getTaskObject() {
        return taskObject;
    }

    public void setTaskObject(ITaskObject taskObject) {
        this.taskObject = taskObject;
    }

    public int getDanceId() {
        return danceId;
    }

    public void setDanceId(int danceId) {
        this.danceId = danceId;
    }

    public boolean isDancing() {
        return danceId > 0;
    }

    public boolean hasEffect() {
        return effectId > 0;
    }

    public int getEffectId() {
        return effectId;
    }

    public void setEffectId(int effectId) {
        this.effectId = effectId;
    }

    public boolean isBeingKicked() {
        return beingKicked;
    }

    public void setBeingKicked(boolean beingKicked) {
        this.beingKicked = beingKicked;
    }

    public Map<String, EntityStatus> getStatuses() {
        return statuses;
    }

    public void setStatuses(Map<String, EntityStatus> statuses) {
        this.statuses = statuses;
    }

    public void reset() {
        statuses = new HashMap<>();
        isWalking = false;
        goal = null;
        next = null;
        instanceId = -1;
        danceId = 0;
        room = null;
        timerManager.reset();
        walkingAllowed = true;
    }

    public void talk(ChatMessageType chatMessageType, String chatMsg, List<Player> receiveMessages) {
        if (receiveMessages == null)
            receiveMessages = new ArrayList<>(room.getEntityManager().getEntities(Player.class));

        List<Player> distinct = new ArrayList<>(new java.util.LinkedHashSet<>(receiveMessages));

        for (Player player : distinct) {
            switch (chatMessageType) {
                case CHAT:
                    player.send(new CHAT(entity.getEntityData().getName(), chatMsg));
                    break;
                case SHOUT:
                    player.send(new SHOUT(entity.getEntityData().getName(), chatMsg));
                    break;
                case WHISPER:
                    player.send(new WHISPER(entity.getEntityData().getName(), chatMsg));
                    break;
            }
        }
    }

    public void talk(ChatMessageType chatMessageType, String chatMsg) {
        talk(chatMessageType, chatMsg, null);
    }

    private int getChatGesture(String chatMsg) {
        chatMsg = chatMsg.toLowerCase();

        if (chatMsg.contains(":)") || chatMsg.contains(":d") || chatMsg.contains("=]")
                || chatMsg.contains("=d") || chatMsg.contains(":>")) {
            return 1;
        }

        if (chatMsg.contains(">:(") || chatMsg.contains(":@"))
            return 2;

        if (chatMsg.contains(":o"))
            return 3;

        if (chatMsg.contains(":(") || chatMsg.contains("=[") || chatMsg.contains(":'(") || chatMsg.contains("='["))
            return 4;

        return 0;
    }

    public void move(int x, int y) {
        if (room == null)
            return;

        if (next != null) {
            Position oldPosition = next.copy();
            position.setX(oldPosition.getX());
            position.setY(oldPosition.getY());
            position.setZ(room.getModel().getTileHeights()[oldPosition.getX()][oldPosition.getY()]);
            needsUpdate = true;
        }

        goal = new Position(x, y);

        RoomTile goalTile = goal.getTile(room);

        if (!RoomTile.isValidTile(room, entity, goal))
            return;

        List<Position> pathList = Pathfinder.findPath(entity, room, position, goal);

        if (pathList == null || pathList.isEmpty())
            return;

        this.pathList = pathList;
        isWalking = true;
    }

    public void kick(boolean allowWalking) {
        this.beingKicked = true;
        Position doorLocation = this.room.getModel().getDoor();

        if (this.position.equals(doorLocation)) {
            this.room.getEntityManager().leaveRoom(this.entity);
            return;
        }

        this.move(doorLocation.getX(), doorLocation.getY());
        this.walkingAllowed = allowWalking;

        if (!this.isWalking) {
            this.room.getEntityManager().leaveRoom(this.entity);
        }
    }

    public void kick() {
        kick(true);
    }

    public void stopWalking() {
        if (!this.isWalking)
            return;

        this.isWalking = false;
        this.pathList.clear();
        this.next = null;
        this.removeStatus("mv");
        this.interactItem();
        this.needsUpdate = true;

        boolean leaveRoom = this.beingKicked;
        Position doorPosition = this.room.getModel().getDoor();

        if (doorPosition.equals(this.position)) {
            leaveRoom = true;
        }

        if (leaveRoom || this.beingKicked) {
            this.room.getEntityManager().leaveRoom(this.entity, true);
            return;
        }
    }

    public void interactItem() {
        RoomTile roomTile = getCurrentTile();

        if (roomTile == null)
            return;

        position.setZ(roomTile.getWalkingHeight());

        Item item = getCurrentItem();

        if (item == null ||
                (!item.getDefinition().getData().isChair()
                && !item.getDefinition().getData().isBed())) {
            if (containsStatus("sit") || containsStatus("lay")) {
                removeStatus("sit");
                removeStatus("lay");
            }
        }

        if (item != null) {
            if (item.getDefinition().getData().isChair()) {
                position.setBodyRotation(item.getPosition().getRotation());
                position.setHeadRotation(item.getPosition().getRotation());
                addStatus("sit", String.valueOf((int) item.getHeight()));
                needsUpdate = true;
            }

            if (item.getDefinition().getData().isBed()) {
                position.setBodyRotation(item.getPosition().getRotation());
                position.setHeadRotation(item.getPosition().getRotation());
                addStatus("lay", String.valueOf((int) item.getHeight()));
                needsUpdate = true;
            }
        }

        this.needsUpdate = true;
    }

    public boolean containsStatus(String statusKey) {
        return statuses.containsKey(statusKey);
    }

    public void addStatus(String key, String value) {
        this.removeStatus(key);
        statuses.put(key, new EntityStatus(value));
    }

    public void removeStatus(String key) {
        statuses.remove(key);
    }

    public void warp(Position targetPosition, boolean instantUpdate) {
        RoomTile oldTile = getCurrentTile();

        if (oldTile != null) {
            oldTile.removeEntity(entity);
        }

        if (next != null) {
            RoomTile nextTile = next.getTile(room);
            if (nextTile != null) {
                nextTile.removeEntity(entity);
            }
        }

        position = targetPosition.copy();
        refreshHeight(targetPosition);

        RoomTile newTile = getCurrentTile();

        if (newTile != null) {
            newTile.addEntity(entity);
        }

        if (instantUpdate && room != null) {
            interactItem();
        }
    }

    public void warp(Position targetPosition) {
        warp(targetPosition, false);
    }

    private void refreshHeight(Position newPosition) {
        Position targetPosition = newPosition != null ? newPosition : position;

        RoomTile oldTile = position.getTile(room);
        RoomTile newTile = targetPosition.getTile(room);

        if (oldTile != null && newTile != null && oldTile.getWalkingHeight() != newTile.getWalkingHeight()) {
            position.setZ(newTile.getWalkingHeight());
            needsUpdate = true;
        }
    }
}
